File: Windows/Prosary.Tests/HebrewLanguageSelectionTests.cs
using Prosary.Localization;
using Prosary.Models;
using Xunit;

namespace Prosary.Tests;

public class HebrewLanguageSelectionTests : IClassFixture<PrayerPackLoaderFixture>
{
    private const string Mission = "he-x-gamliel";

    public HebrewLanguageSelectionTests(PrayerPackLoaderFixture _) { }

    private static void WithOrder(string[] order, Action check)
    {
        var previous = AppSettings.LanguageFallbackOrder.ToArray();
        try { AppSettings.SetLanguageFallbackOrder(order); check(); }
        finally { AppSettings.SetLanguageFallbackOrder(previous); }
    }

    [Fact]
    public void EnteringHebrewUsesFirstSavedTraditionFromOtherLanguagesAndAppSetting()
    {
        foreach (var order in new[] { new[] { Mission, "arc", "he" }, new[] { "he", "arc", Mission } })
            WithOrder(order, () =>
            {
                foreach (var current in LanguageCatalog.All.Select(option => option.Code).Where(code => code != "he" && code != Mission).Append(""))
                    Assert.Equal(order[0], LanguageCatalog.SelectingLanguage("he", current));
                Assert.Equal(order, AppSettings.LanguageFallbackOrder);
            });
    }

    [Fact]
    public void AlreadySelectedHebrewTraditionRemainsExplicitInEitherOrder()
    {
        foreach (var order in new[] { new[] { Mission, "arc", "he" }, new[] { "he", "arc", Mission } })
            WithOrder(order, () =>
            {
                foreach (var current in new[] { "he", Mission })
                    Assert.Equal(current, LanguageCatalog.SelectingLanguage("he", current));
            });
    }

    [Fact]
    public void LeavingHebrewAndReturningUsesPriorityAgain()
    {
        foreach (var (order, previous) in new[] { (new[] { Mission, "arc", "he" }, "he"), (new[] { "he", "arc", Mission }, Mission) })
            WithOrder(order, () =>
            {
                var english = LanguageCatalog.SelectingLanguage("en", previous);
                Assert.Equal("en", english);
                Assert.Equal(order[0], LanguageCatalog.SelectingLanguage("he", english));
            });
    }

    [Fact]
    public void OtherLanguageAndEmptyChoicesArePassedThrough() => WithOrder([Mission, "arc", "he"], () =>
    {
        foreach (var next in LanguageCatalog.All.Select(option => option.Code).Where(code => code != "he").Concat(["", "unknown"]))
        foreach (var current in new[] { "he", Mission, "en", "" })
            Assert.Equal(next, LanguageCatalog.SelectingLanguage(next, current));
    });

    [Fact]
    public void EmptyAndDamagedSavedOrdersUseOnlyKnownDistinctTraditions()
    {
        foreach (var (order, expected) in new[] {
            (Array.Empty<string>(), "he"), (new[] { "unknown" }, "he"),
            (new[] { "unknown", LanguageCatalog.VicariateContentCode, Mission, Mission, "arc", "he" }, Mission),
            (new[] { "unknown", "he", "he", "arc", Mission }, "he"),
        }) WithOrder(order, () =>
        {
            Assert.Equal(expected, LanguageCatalog.SelectingLanguage("he", ""));
            Assert.Single(LanguageCatalog.FallbackOrder, code => code == expected);
            Assert.DoesNotContain("unknown", LanguageCatalog.FallbackOrder);
            Assert.DoesNotContain(LanguageCatalog.VicariateContentCode, LanguageCatalog.FallbackOrder);
        });
    }

    [Fact]
    public void SelectedMissionCodeReachesItsRealPrayerTextAndAppDefault() => WithOrder([Mission, "arc", "he"], () =>
    {
        var previousDefault = AppSettings.DefaultLanguageCode;
        try
        {
            var prayer = BasicPrayerCatalog.Prayer("ourFather")!;
            var expected = PrayerTranslations.ByLanguage[Mission][PrayerKey.PaterNoster];
            var selected = LanguageCatalog.SelectingLanguage("he", "en");
            var step = BasicPrayerCatalog.Step(prayer, selected);
            Assert.Equal(expected, step.Body);
            Assert.Equal("תפילת האדון", step.Title);
            Assert.NotEqual(step.Body, BasicPrayerCatalog.Step(prayer, "he").Body);
            AppSettings.SetDefaultLanguageCode(selected);
            Assert.Equal(expected, BasicPrayerCatalog.Step(prayer, "").Body);
            Assert.Equal(Mission, LanguageCatalog.Resolve("").Code);
        }
        finally { AppSettings.SetDefaultLanguageCode(previousDefault); }
    });
}

File: Android/app/src/test/java/com/dkaluta/prosary/models/HebrewLanguageSelectionTests.kt
package com.dkaluta.prosary.models

import com.dkaluta.prosary.content.PrayerKey
import com.dkaluta.prosary.content.prayerTranslationsHebrewGamaliel
import com.dkaluta.prosary.content.prayerpack.PrayerPackStore
import java.io.File
import org.junit.Assert.*
import org.junit.BeforeClass
import org.junit.Test

class HebrewLanguageSelectionTests {
    private val mission = "he-x-gamliel"

    companion object {
        @BeforeClass @JvmStatic fun loadPacks() {
            PrayerPackStore.initialize { name ->
                File("src/main/assets/$name.prosaryprayer").takeIf(File::exists)?.inputStream()
            }
        }
    }

    private fun withOrder(order: List<String>, check: () -> Unit) {
        val previous = AppSettings.languageFallbackOrder
        try {
            AppSettings.setLanguageFallbackOrder(order)
            check()
        } finally { AppSettings.setLanguageFallbackOrder(previous) }
    }

    @Test fun enteringHebrewUsesFirstSavedTraditionFromOtherLanguagesAndAppSetting() {
        for (order in listOf(listOf(mission, "arc", "he"), listOf("he", "arc", mission))) withOrder(order) {
            for (current in LanguageCatalog.all.map { it.code }.filter { it != "he" && it != mission } + "") {
                assertEquals(current, order.first(), LanguageCatalog.selectingLanguage("he", current))
            }
            assertEquals(order, AppSettings.languageFallbackOrder)
        }
    }

    @Test fun alreadySelectedHebrewTraditionRemainsExplicitInEitherOrder() {
        for (order in listOf(listOf(mission, "arc", "he"), listOf("he", "arc", mission))) withOrder(order) {
            for (current in listOf("he", mission)) assertEquals(current, LanguageCatalog.selectingLanguage("he", current))
        }
    }

    @Test fun leavingHebrewAndReturningUsesPriorityAgain() {
        for ((order, previous) in listOf(listOf(mission, "arc", "he") to "he", listOf("he", "arc", mission) to mission)) withOrder(order) {
            val english = LanguageCatalog.selectingLanguage("en", previous)
            assertEquals("en", english)
            assertEquals(order.first(), LanguageCatalog.selectingLanguage("he", english))
        }
    }

    @Test fun otherLanguageAndEmptyChoicesArePassedThrough() = withOrder(listOf(mission, "arc", "he")) {
        for (next in LanguageCatalog.all.map { it.code }.filter { it != "he" } + listOf("", "unknown")) {
            for (current in listOf("he", mission, "en", "")) assertEquals(next, LanguageCatalog.selectingLanguage(next, current))
        }
    }

    @Test fun emptyAndDamagedSavedOrdersUseOnlyKnownDistinctTraditions() {
        for ((order, expected) in listOf(
            emptyList<String>() to "he", listOf("unknown") to "he",
            listOf("unknown", LanguageCatalog.hebrewVicariateContentCode, mission, mission, "arc", "he") to mission,
            listOf("unknown", "he", "he", "arc", mission) to "he",
        )) withOrder(order) {
            assertEquals(expected, LanguageCatalog.selectingLanguage("he", ""))
            assertEquals(1, LanguageCatalog.fallbackOrder.count { it == expected })
            assertFalse("unknown" in LanguageCatalog.fallbackOrder)
            assertFalse(LanguageCatalog.hebrewVicariateContentCode in LanguageCatalog.fallbackOrder)
        }
    }

    @Test fun selectedMissionCodeReachesItsRealPrayerTextAndAppDefault() = withOrder(listOf(mission, "arc", "he")) {
        val previousDefault = AppSettings.defaultLanguageCode
        try {
            val prayer = requireNotNull(BasicPrayerCatalog.prayer("ourFather"))
            val expected = prayerTranslationsHebrewGamaliel.getValue(PrayerKey.PaterNoster)
            val selected = LanguageCatalog.selectingLanguage("he", "en")
            val step = BasicPrayerCatalog.step(prayer, selected)
            assertEquals(expected, step.body)
            assertEquals("תפילת האדון", step.title)
            assertNotEquals(step.body, BasicPrayerCatalog.step(prayer, "he").body)
            AppSettings.setDefaultLanguageCode(selected)
            assertEquals(expected, BasicPrayerCatalog.step(prayer, "").body)
            assertEquals(mission, LanguageCatalog.resolve("").code)
        } finally { AppSettings.setDefaultLanguageCode(previousDefault) }
    }
}
